File: External-Tutorial-SubComponentBuilder/src/main/java/zorail/rohan/com/subcomponent/component/AppComponent.java
package zorail.rohan.com.subcomponent.component;

import android.app.Application;

import java.util.Map;

import javax.inject.Provider;
import javax.inject.Singleton;

import dagger.Component;
import zorail.rohan.com.subcomponent.Binder.ApplicationBinder;
import zorail.rohan.com.subcomponent.SubComponentBuilder;
import zorail.rohan.com.subcomponent.module.AppModule;

/**
 * Created by zorail on 22-May-17.
 */
@Singleton
@Component(modules ={ AppModule.class, ApplicationBinder.class})
public interface AppComponent {
   Map<Class<?>,Provider<SubComponentBuilder>> subcomponentBuilders();
}

File: Tutorial6-1ComponentFactory/src/main/java/com/smarttoolfactory/tutorial6_1componentfactory/di/AppComponent.kt
package com.smarttoolfactory.tutorial6_1componentfactory.di

import android.app.Application
import android.content.SharedPreferences
import com.smarttoolfactory.tutorial6_1componentfactory.MainActivity
import dagger.BindsInstance
import dagger.Component

@Component(modules = [AppModule::class])
interface AppComponent {

    fun inject(mainActivity: MainActivity)

    @Component.Factory
    interface Factory {

        fun create(@BindsInstance application: Application,
                   @BindsInstance sharedPreferences: SharedPreferences): AppComponent
    }

}